Add new docs for OTP authentication (#41748)
mdn/content: files/en-us/web/security/authentication/otp/index.md

```plain
Your verification code is 123456.

@www.example.com #123456
```

### HTML markup

In order for a website to find the associated SMS and propose the OTP code for autocompletion, provide an {{HTMLElement("input")}} element with the `autocomplete=one-time-code` attribute value set.

```html
<form action="/verify-otp" method="POST">
  <input
    required
    type="text"
    autocomplete="one-time-code"
    inputmode="numeric"
    maxlength="6"
    pattern="\d{6}" />
  <input type="submit" />
</form>
```

### WebOTP API

In Chromium-based browsers, the [WebOTP API](/en-US/docs/Web/API/WebOTP_API) is available additionally as an extension of the [Credential Management API](/en-US/docs/Web/API/Credential_Management_API). Unless you need programmatic access to the {{domxref("OTPCredential")}}, you don't need to use this API. Formatting the SMS in the standardized format and using `autocomplete=one-time-code` should be enough for autocomplete to work across browsers.

## TOTP flows (authenticator app)

> The time-based one-time password (TOTP) algorithm is specified in [RFC 6238](https://www.rfc-editor.org/rfc/rfc6238). It creates one-time codes that are 6 digits long and which are only valid for a limited amount of time (usually 30 seconds). It is similar to other OTP systems but implements time-based validity and automatic invalidation by design.

Reviewer: Also and maybe more importantly: with TOTP the code is not sent by the server. Any time the code is sent from the server to the user, there is the possibility that it could be intercepted. So in this sense TOTP is intrinsically more secure.

Reviewer: Noted in "Weaknesses" in https://github.com/mdn/content/pull/41748/commits/513025271e8f15df1e487f02e42ef681e4933bea.

In order for TOTP to work, the following steps will need to happen:

1. The website generates and shares a secret seed (which can be made accessible using a QR code).
2. The user generates a TOTP token based on the secret seed and the current time (authenticator apps can help with this).
3. The user enters the TOTP token on the website to authenticate.

### TOTP secret sharing

> To setup TOTP, a website needs to share a secret key with the user. A TOTP secret key is a randomly generated [Base32](https://en.wikipedia.org/wiki/Base32) encoded seed. For example, the following command generates a 32-character Base32 random string::

Reviewer: ```suggestion
To set up TOTP, a website needs to share a secret key with the user. A TOTP secret key is a randomly generated [Base32](https://en.wikipedia.org/wiki/Base32) encoded value. For example, the following command generates a 32-character Base32 random string::
```

...but also I think we need more context than "the following command" - I guess it is a bash command, but you don't say that. And what is it actually doing? I don't want to paste random bash commands into my website code without some idea of what the bits of them are.

And talking about flows, it's better to always say _who_ needs to do _what_.

> ```plain

Reviewer: ````suggestion
```bash
````

?

LC_ALL=C tr -dc 'A-Z2-7' </dev/urandom | head -c 32; echo
```

> This secret can be provided to the user in the form of a QR code, however, if you want to avoid QR codes, all you really need is an URI in the `otpauth` URI format as defined in this [IETF draft](https://www.ietf.org/archive/id/draft-linuxgemini-otpauth-uri-00.html).

Reviewer: I don't fully understand this, and again I think spelling it out is helpful. For example:

- the user asks to create an account
- the website generates a secret, and then generates a QR code from that secret (by doing XYZ)
- the user ...(does something with the QR code)

I don't really understand what the website does with this `otpauth` URI. How does this enable sharing the secret value.

```plain
otpauth://totp/ExampleWebsite?secret=MQCHJLS6FJXT2BGQJ6QMG3WCAVUC2HJZ
```

The URI is an URL query string and it has a few parameters (support in authenticator applications may vary). The most interesting parameters are:

- `secret`
  - : The shared secret encoded in `base32`.
- `algorithm` {{optional_inline}}
  - : The algorithm used for the TOTP tokens. Possible values: `sha1` (default), `sha256`, `sha512`.
- `digits` {{optional_inline}}
  - : The number of digits that should be outputted for the TOTP token. 6 digits is the default value.
- `period` {{optional_inline}}
  - : The refresh period in seconds for the TOTP tokens.

If you'd like to use QR codes, you can use [`qrencode`](https://fukuchi.org/works/qrencode/) or similar libraries and create an image from the command line, for example:

```plain
qrencode -t PNG -o qr-totp.png 'otpauth://totp/ExampleWebsite?secret=MQCHJLS6FJXT2BGQJ6QMG3WCAVUC2HJZ'
```

### Generating TOTP codes

With the shared secret key and the current time as inputs, you can now produce a unique, time-based code. This usually gets done using a cryptographic algorithm, typically `HMAC-SHA1` or `HMAC-SHA256`.

> Several apps, like [Ente Auth](https://ente.io/auth/), [2FAS](https://2fas.com) or [Microsoft Authenticator](https://www.microsoft.com/en-US/security/mobile-authenticator-app), allow you to store secrets you've received from various websites or services. You don't necessarily need to use an authenticator app, but if you do, choose one that is trustworthy and secure. For example, there are several TOTP token generator tools online, but you might not know what they are doing with your shared secrets. If you use online services to quickly generate TOTP tokens, at least consider to not provide details for which service the shared secret is intended for.

Reviewer: "you" here seems to refer to the user, not the web developer. Which I think is wrong: we should consistently address the web developer.

What are the security implications of the shared secret? What if an attacker gets access to the secret? In the preamble, we say "Unlike a password, there's no long-term secret to remember or to store securely" (my words actually!) Is this incorrect?

Reviewer: Relatedly it would be good to talk about attacks on (T)OTP. Earlier on we mention phishing but it's kind of implied that this is only a concern for certain versions of email OTP, but I think all OTP variants are vulnerable to phishing: https://developer.mozilla.org/en-US/docs/Web/Security/Attacks/Phishing#multi-factor_authentication.

Somewhere too we should talk about pros and cons. One con I can think of is that users have to install an authenticator app, which is extra friction and another component that we have to trust (I guess?).

## OTP recommendations

OTP, and in particular TOTP, can be useful as an additional (2FA) authentication method and for confirming user intentions, for example when making a payment. For general authentication purposes, much stronger mechanism, such as [WebAuthn](/en-US/docs/Web/API/Web_Authentication_API), should be considered. If you implement one-time passwords, consider the following recommendations:

- Bind the (T)OTP to a single session or transaction only.
- Use TOTP instead of your own OTP algorithm.
- If you implement your own OTP:
  - Use 6 or more digits.
  - Set short expiry (ideally ≤5 minutes; 30–120 seconds for stronger protection).
  - Invalidate OTPs after use.
- If you use email:

>   - Ensure users receive the OTP emails quickly.

Reviewer: How can a website do this?

  - Prefer one-time authorization codes over providing clickable links.
- If you use SMS:
  - Format messages in origin-bound standard format.
  - Use `autocomplete=one-time-code`.

>   - Monitor and mitigate SIM swap and interception risks.

Reviewer: I wonder if it's worth being stronger here and just straight out un-recommending SMS, rather than giving good practice recs.

- Carriers can also recycle phone numbers to new users after an account is closed.

Because of this, you should not use SMS OTP on its own to establish new sessions or for general authentication. Instead, if at all, only use it as a second factor or for confirming intentions (e.g., payments).

### Autocompleting SMS codes

To make it easier for users to enter SMS codes into a site, and to reduce the likelihood of phishing attacks, the [standard for origin-bound one-time codes delivered via SMS](https://wicg.github.io/sms-one-time-codes/) enables websites to support autocomplete for one-time code values.

To enable this, you need to format the SMS message like this:

```plain
Your verification code is 123456.

@www.example.com #123456
```

Then in your site's login form, provide an {{HTMLElement("input")}} element with the `autocomplete=one-time-code` attribute value set.

```html
<form action="/verify-otp" method="POST">
  <input
    required
    type="text"
    autocomplete="one-time-code"
    inputmode="numeric"
    maxlength="6"
    pattern="\d{6}" />
  <input type="submit" />
</form>
```

> The browser will automatically extract the code from the SMS, and if the origin given in the message matches the origin of the login form, will autofill the `<input>` element with the code.

Author: ```suggestion
The browser will automatically extract the code from the SMS, and if the origin given in the message matches the origin of the login form, it will autofill the `<input>` element with the code.
```

### WebOTP API

The [WebOTP API](/en-US/docs/Web/API/WebOTP_API) gives websites programmatic access to one-time codes delivered over SMS. However, it doesn't have good cross-browser support, and unless you need programmatic access to the code, you don't need to use this API. Using the standardized format and `autocomplete=one-time-code` should be enough for autocomplete to work across browsers.

## TOTP

> With time-based one-time passwords, the website does not send the sign-in code to the user. Instead, the website and the user are able to generate the same code independently of each other, based on the current time and a shared secret. To generate the code, the user must install an app on their device: this is called an _authenticator app_.

Author: Yeah, I agree, let's not get into this then.
